### ISSUE TITLE
Support include/skip directive for recursive fragments

### ISSUE THREAD

**Issue by bgnorlov** [facebook/relay#524]
Hi guys! I have the big problem with relay. I tried to create a simple tree component and got stack overflow error in this part of Relay's code
```javascript
  RelayQueryNode.prototype.hasDeferredDescendant = function hasDeferredDescendant() {
    var hasDeferredDescendant = this.__hasDeferredDescendant__;
    if (hasDeferredDescendant == null) {
      hasDeferredDescendant = !this.isScalar() && this.getChildren().some(function (child) {
        return child.hasDeferredDescendant();
      });
      this.__hasDeferredDescendant__ = hasDeferredDescendant;
    }
    return hasDeferredDescendant;
  };
```
My code is simple tree component (demo on [heroku](https://relay-sample-tree.herokuapp.com), full source code [here](https://github.com/bgnorlov/relay-sample-tree) ):

```javascript
import React from 'react';
import Relay from 'react-relay';
class TreeNode extends React.Component {
  handleClick() {
    this.props.relay.setVariables({
      expanded: !this.props.relay.variables.expanded
    });
  }
  render() {
    var node = this.props.node;
    var variables = this.props.relay.variables;
    return (
      <div className="node">
        <div className="title" onClick={this.handleClick.bind(this)}>{node.title}</div>
        <div className="children" style={{paddingLeft: 20}}>
          {variables.expanded && node.children && node.children.edges.map((edge)=> {
            return <TreeNodeContainer node={edge.node} key={edge.node.id}/>
          })}
        </div>
      </div>
    )
  }
}
const TreeNodeContainer = Relay.createContainer(TreeNode, {
  initialVariables: {
    expanded: false
  },
  fragments: {
    node: () => {
      return Relay.QL`
      fragment on TreeNode {
        id,
        title,
        children(first: 10) @include(if: $expanded) {
          edges {
            node {
              id,
              ${TreeNodeContainer.getFragment('node')}
            }
          }
        }
      }
    `
    }
  },
});
export default TreeNodeContainer;
```

**Comment by nodkz** [comment]
I suppose that `Relay.QL` is isolated from data/state when convert the graphql-template to a js-code. So I think that `@include(if: $expanded)` works only when query is already converted to JS. According to this, you get `Stack overflow`.  

Try to move `children` to another component. And and check `if $expanded` like normal js code in arrow function. May be it help.

PS. It would be great I FB-relay team add example how to construct tree with Relay.

**Comment by steveluscher** [comment]
We support recursive queries, except via directives (like `@include`). I'll mark this as an enhancement.

In the meantime, you can use `.if(…)` on a fragment to stop the recursion, as demonstrated in this Stack Overflow answer: http://stackoverflow.com/a/32508427/802047

```js
TreeNodeList = Relay.createContainer(TreeNodeList, {
  fragments: {
    list: () => Relay.QL`
      fragment on TreeNode {
        children(first: 10) {
          edges {
            node {
              id
              ${TreeNode.getFragment('node')}
            }
          }
        }
      }
    `,
  },
});

TreeNode = Relay.createContainer(TreeNode, {
  initialVariables: {
    expanded: false,
  },
  fragments: {
    node: variables => Relay.QL`
      fragment on TreeNode {
        id
        title
        ${TreeNodeList.getFragment('list').if(variables.expanded)}
      }
    `,
  },
});
```

**Comment by josephsavona** [comment]
merged into #536
